### PR TITLE
Only call the cleanup methods for existing state keys.

### LINKED ISSUE #3481: Closing entity details by the search box X does not return to the canvas
To reproduce:
- Add a charm to the canvas
- In the inspector, click the 'charm details' link
- Click the 'X' in the search box, the charm details will still be rendered.
- Click the search box again,
- Click the 'X', it will now return to the canvas.

### PR DESCRIPTION
**Pull request by hatched**
Fixes #3481

Prior to this change, calls to changeState where you've set a key to null would always execute the cleanup method for that key. This meant that the cleanup methods had to be idempotent and was a minor performance issue. State will now only call cleanup methods if the state key previously existed.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

Make sure that the GUI works, that the back button works (keep in mind that there are extraneous states that make the back button functionality clumsy (#3085 related))
Repeat the steps in #3481 
Also open the inspector and show a post deployment panel, then close it.

File: jujugui/static/gui/src/app/state/test-state.js [hunk 1]
```diff
@@ -1359,9 +1359,16 @@ describe('State', () => {
         location: {href: '/u/hatch/staging/i/applications/inspector/ghost'}
       });
       const pushStub = sinon.stub(state, '_pushState');
-      state.dispatch();
+      state.bootstrap();
```
Comment (hatched) [inline jujugui/static/gui/src/app/state/test-state.js]: These should have been bootstrap all along. If it was, this issue would have been caught much sooner.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8508/

**Comment by mitechie** [comment]
QA ok, there's a few design issues and I've filed a bug for that but this fixes up the functionality and with this, the logout link, and the login to charmstore buttons we're in solid shape.

**Comment by mitechie** [comment]
👍

**Comment by hatched** [comment]
Thanks @mitechie 

:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**